Nginx1.26.patch and fix raise count does not increase under the TCP check

mofantor opened the pull request:
+ 新增ng1.26.3补丁
+ 修改README描述布局（使用markdown）
+ 修复在TCP健康检查方法下，Rise Count不增加的问题

**原因分析**

在tcp检查下，默认keepalive的请求是1，但http等方式的请求再发送请求后，都会增加connection的requests数量，但在tcp连接方式中没有增加，导致tcp连接不会被关闭
```c
// ngx_http_upstream_check_send_handler
  if (ctx->send.pos == ctx->send.last) {
      ngx_log_debug0(NGX_LOG_DEBUG_HTTP, c->log, 0, "http check send done.");
      peer->state = NGX_HTTP_CHECK_SEND_DONE;
      c->requests++; // http等检查方式中，增加当前连接的请求数量
  }

//ngx_http_upstream_check_clean_event
        if (c->error == 0 &&
            cf->need_keepalive &&
            (c->requests < ucscf->check_keepalive_requests)) // 这里的c->requests永远是0，没有增加，导致不会进入关闭连接
        {
            c->write->handler = ngx_http_upstream_check_dummy_handler;
            c->read->handler = ngx_http_upstream_check_discard_handler;
        } else {
            ngx_close_connection(c);
            peer->pc.connection = NULL;
        }
```
所以修复措施就是在tcp连接检查没有问题后，增加一次requsets
```c
// ngx_http_upstream_check_peek_handler
    if (ngx_http_upstream_check_peek_one_byte(c) == NGX_OK) {
        ngx_http_upstream_check_status_update(peer, 1);
        c->requests++;   // 增加requests数量
    } else {
        c->error = 1;
        ngx_http_upstream_check_status_update(peer, 0);
    }
```

Commit:
Patch for nginx 1.26.0+

Commit:
update README

Commit:
Fix the issue where the raise count does not increase under the TCP check method.

Commit:
Fix the potential overflow issue of rise_count and fall_count, which may lead to service unavailability.

noeldamps commented:
@mofantor Saw your patch that reference to not closing tcp connections. Is it possible thats the same error, that I got in my Pull Request pr #281 ?

mofantor commented:
I think the HTTPS health check method you submitted doesn’t have this
issue.
In the TCP connection, the request count c->requests isn’t incremented
after sending the request,
which causes the incorrect evaluation in
ngx_http_upstream_check_clean_event: c->requests <
ucscf->check_keepalive_requests.
However, your HTTPS implementation references
ngx_http_upstream_check_send_handler, where the requests count is correctly
incremented.
This ensures that connections exceeding or matching
check_keepalive_requests are properly released in
ngx_http_upstream_check_clean_event

On Fri, 21 Mar 2025 at 17:48, Noel Damps ***@***.***> wrote:

> @mofantor <https://github.com/mofantor> Saw your patch that reference to
> not closing tcp connections. Is it possible thats the same error, that I
> got in my Pull Request pr #281
> <https://github.com/yaoweibin/nginx_upstream_check_module/pull/281> ?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/yaoweibin/nginx_upstream_check_module/pull/282#issuecomment-2742849963>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/A7T5WGYW3RHOB7SB24U4XWT2VPN5XAVCNFSM6AAAAABZM6YJFOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDONBSHA2DSOJWGM>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
> [image: noeldamps]*noeldamps* left a comment
> (yaoweibin/nginx_upstream_check_module#282)
> <https://github.com/yaoweibin/nginx_upstream_check_module/pull/282#issuecomment-2742849963>
>
> @mofantor <https://github.com/mofantor> Saw your patch that reference to
> not closing tcp connections. Is it possible thats the same error, that I
> got in my Pull Request pr #281
> <https://github.com/yaoweibin/nginx_upstream_check_module/pull/281> ?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/yaoweibin/nginx_upstream_check_module/pull/282#issuecomment-2742849963>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/A7T5WGYW3RHOB7SB24U4XWT2VPN5XAVCNFSM6AAAAABZM6YJFOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDONBSHA2DSOJWGM>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

mofantor committed:
Fix the issue where the rise count does not increase under the TCP check method

mofantor committed:
Patch for nginx 1.28.1+

mofantor committed:
Add patch file for version 1.28.1

mofantor committed:
Patch for nginx 1.30.0+

Akvicor commented:
重写 [https://github.com/Akvicor/nginx-healthcheck-module](https://github.com/Akvicor/nginx-healthcheck-module)

Commit:
Add zone+resolve per-IP health checking (V2) for nginx 1.30+

mofantor committed:
Merge pull request #3 from dnj12345/master

Add zone+resolve per-IP health checking (V2) for nginx 1.30+

Commit:
Security hardening for zone+resolve V2 health check code

- Cap check_resolve_max_addrs at 64 to prevent memory exhaustion at config parse time
- Validate socklen against sizeof(sockaddr_storage) in get_resolve_peer before memcpy
- Guard NULL name pointer in get_resolve_peer with safe_name fallback for log calls
- Guard NULL name in init_shm_peer on non-atomic platforms to prevent crash at startup
- Use ngx_atomic_fetch_add acquire-load in begin_handler for correct ARM memory ordering

dnj12345 committed:
Merge branch 'mofantor:master' into master

mofantor committed:
Merge pull request #4 from dnj12345/master

Security hardening for zone+resolve V2 health check code
